query/sources/impl/src/test/groovy/org/codice/ddf/admin/sources/csw/discover/DiscoverCswSourceSpec.groovy

```diff
@@ -0,0 +1,271 @@
+        config.outputSchema() == CswSourceUtils.CSW_2_0_2_OUTPUT_SCHEMA
+        config.spatialOperator() == NO_FILTER
+        config.credentials().password() == FLAG_PASSWORD
+    }
+
+    def 'Successfully discover GMD CSW federation profile using hostname and port'() {
+        setup:
+        discoverCsw.setCswSourceUtils(prepareCswSourceUtils(200, gmdCswResponse, true))
+        discoverCsw.setValue(getBaseDiscoverByAddressArgs())
+
+        when:
+        def report = discoverCsw.getValue()
+        def config = (CswSourceConfigurationField) report.result()
+
+        then:
+        config.endpointUrl() == TEST_CSW_URL
+        config.cswProfile() == CswProfile.GmdCswFederatedSource.GMD_CSW_ISO_FEDERATED_SOURCE
+        config.outputSchema() == CswSourceUtils.GMD_OUTPUT_SCHEMA
+        config.spatialOperator() == NO_FILTER
+        config.credentials().password() == FLAG_PASSWORD
+    }
+
+    def 'Unknown endpoint error with bad HTTP status'() {
+        setup:
+        discoverCsw.setCswSourceUtils(prepareCswSourceUtils(500, gmdCswResponse, true))
+        discoverCsw.setValue(getBaseDiscoverByUrlArgs(TEST_CSW_URL))
+
+        when:
+        def report = discoverCsw.getValue()
+
+        then:
+        report.messages().size() == 1
+        report.messages()[0].getCode() == DefaultMessages.UNKNOWN_ENDPOINT
+        report.messages()[0].getPath() == [DiscoverCswSource.FIELD_NAME]
+    }
+
+    def 'Unknown endpoint error with unrecognized response when using URL'() {
+        setup:
+        discoverCsw.setCswSourceUtils(prepareCswSourceUtils(200, badResponseBody, true))
+        discoverCsw.setValue(getBaseDiscoverByUrlArgs(TEST_CSW_URL))
+
+        when:
+        def report = discoverCsw.getValue()
+
+        then:
+        report.messages().size() == 1
+        report.messages()[0].getCode() == DefaultMessages.UNKNOWN_ENDPOINT
+        report.messages()[0].getPath() == [DiscoverCswSource.FIELD_NAME]
+    }
+
+    def 'Unknown endpoint error with unrecognized response when using hostname+port'() {
+        setup:
+        discoverCsw.setCswSourceUtils(prepareCswSourceUtils(200, badResponseBody, true))
+        discoverCsw.setValue(getBaseDiscoverByAddressArgs())
+
+        when:
+        def report = discoverCsw.getValue()
+
+        then:
+        report.messages().size() == 1
+        report.messages()[0].getCode() == DefaultMessages.UNKNOWN_ENDPOINT
+        report.messages()[0].getPath() == [DiscoverCswSource.FIELD_NAME]
+    }
+
+    def 'Cannot connect if errors from discover url from host'() {
+        setup:
+        discoverCsw.setCswSourceUtils(prepareCswSourceUtils(200, badResponseBody, false))
+        discoverCsw.setValue(getBaseDiscoverByAddressArgs())
+
+        when:
+        def report = discoverCsw.getValue()
+
+        then:
+        report.messages().size() == 1
+        report.messages()[0].getCode() == DefaultMessages.CANNOT_CONNECT
```

Comment:
The reason it is cannot connect is because we do not know which of the pre formatted URLs failed.